cpp/src/arrow/record_batch.cc

```diff
@@ -334,42 +347,51 @@ Result<std::shared_ptr<Tensor>> RecordBatch::ToTensor(MemoryPool* pool) const {
     result_type = result_field->type();
   }
 
+  // Check if result_type is signed or unsigned integer and null_to_nan is set to true
+  // Then all columns should be promoted to float type
+  if (is_integer(result_type->id()) && null_to_nan) {
+    ARROW_ASSIGN_OR_RAISE(
+        result_field,
+        result_field->MergeWith(field(result_field->name(), float16()), options));
```

Comment:
One problem with using float16 here is that for a combo of (int8, float16) the merged type is still float16 (I think, didn't check), and we can't cast (yet) to float16. 
So maybe use float32 here, so the minimum resulting float type is float32 (and also add a test case for it)

Reply:
Sorry, thought I used `float32` but must have been too late for me yesterday 😊 Will correct!

Reply:
In the current tests there is a case for `int32 + float32 -> float64` and `int32 -> float64`. Should I also add `int8` with nulls?

Reply:
I was going to suggest it would be good to add a test that covers the fact that we need to use float32 here, and not accidentally use float16 (i.e. you changed this now, but the tests were not failing because of it or you didn't update any test). 

But, for the case of (int8, float16), that's actually not supported anyway, as you already raise an error for that combo earlier in the code because of not supporting float16 with mixed types

Reply:
Ah, but of course you can also write the test without float16 in the input, and just have int8 that need to be upcasted to float32 and not float16. So on your

> Should I also add int8 with nulls?

Yes, let's do that. As that is the case that would have been silently giving garbage results when using float16 instead of float32

Reply:
Will do 👍

Reply:
https://github.com/apache/arrow/pull/40803/commits/2aa36801609940c895fd178d328a2b6be3423a29